llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -1971,14 +1971,9 @@ class GeneratedRTChecks {
           // count. Assume that the outer loop executes at least twice.
           unsigned BestTripCount = 2;
 
-          // If exact trip count is known use that.
-          if (unsigned SmallTC = SE->getSmallConstantTripCount(OuterLoop))
-            BestTripCount = SmallTC;
-          else if (LoopVectorizeWithBlockFrequency) {
-            // Else use profile data if available.
-            if (auto EstimatedTC = getLoopEstimatedTripCount(OuterLoop))
-              BestTripCount = *EstimatedTC;
-          }
+          // Get the best known TC estimate.
+          if (auto EstimatedTC = getSmallBestKnownTC(*SE, OuterLoop))
```

Comment:
I'm not sure if this is what we want because getSmallBestKnownTC will fall back on the small constant maximum trip count if it fails to calculate the exact one. For example, the max could be 1024 but the actual runtime trip count might only be 16. That's why the code was written this way.

I like the idea of reusing code though. Perhaps you could add a new flag to `getSmallBestKnownTC` indicating whether we can use the constant max or not? You could make the default value true, and pass in false here, i.e.

```
bool getSmallBestKnownTC(ScalarEvolution *, Loop *, bool CanUseConstantMax = true) {
...
}

...
unsigned BestTripCount = 2;
if (auto EstimatedTC = getSmallBestKnownTC(*SE, OuterLoop, /* CanUseConstantMax=*/false)) {
...
```

Any thoughts @fhahn?

Reply:
Yes, returning the max trip count here would probably severely overestimate the trip count. I *think* some of the existing callers probably should also be updated to avoid using the constant max, especially if it is very large (.e.g -1)

Would be good to have a test that shows the difference between using ` /* CanUseConstantMax=*/false` and ` /* CanUseConstantMax=*/true`